### PR TITLE
add service controller for identity registry support

### PR DESCRIPTION
**Pull request by crazytan**
**What this PR does / why we need it**:
It adds a service controller, that is similar to secret controller and listens for events related to Service objects. Therefore another goroutine is started in CA which does nothing currently. `service_test.go` demonstrates how a very simple registry should work with the controller.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/istio/old_auth_repo/issues/293

**Special notes for your reviewer**:
I also removed unreachable code at the end of `main()`.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: security/pkg/pki/ca/controller/service_test.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	controller := NewServiceController(client.CoreV1(), namespace, r.simpleAdd, r.simpleDelete, r.simpleUpdate)
+	for id, c := range testCases {
+		r.m = make(map[string]bool)
+
+		if c.toAdd != nil {
+			controller.serviceAdded(c.toAdd)
+		}
+		if c.toDelete != nil {
+			controller.serviceDeleted(c.toDelete)
+
```
Comment (lookuptable) [inline security/pkg/pki/ca/controller/service_test.go]: Extra blank line

File: security/cmd/istio_ca/main.go [hunk 2]
```diff
@@ -141,6 +142,12 @@ func runCA() {
 	stopCh := make(chan struct{})
 	sc.Run(stopCh)
 
+	serviceController := controller.NewServiceController(cs.CoreV1(), opts.namespace,
+		func(*v1.Service) {},
```
Comment (lookuptable) [inline security/cmd/istio_ca/main.go]: Is this WIP as all the handlers are empty?

Also it'd be strange if eventually all the handlers are defined here in `main.go`.

File: security/cmd/istio_ca/main.go [hunk 3]
```diff
@@ -149,9 +156,7 @@ func runCA() {
 	}
 
 	glog.Info("Istio CA has started")
-
-	<-stopCh
-	glog.Warning("Istio CA has stopped")
+	select {} // wait forever
```
Comment (lookuptable) [inline security/cmd/istio_ca/main.go]: Just curious why this change is needed?

File: security/cmd/istio_ca/main.go (hunk 2 above)
Comment (crazytan) [reply security/cmd/istio_ca/main.go]: I want to submit controller and registry in two PRs so I added an empty controller here.

Yes it is. In next PR registry will be created in its own package and the handlers are defined there.

File: security/pkg/pki/ca/controller/secret.go [hunk 4]
```diff
@@ -113,7 +113,7 @@ func NewSecretController(ca ca.CertificateAuthority, core corev1.CoreV1Interface
 	return c
 }
 
-// Run starts the SecretController until stopCh is closed.
+// Run starts the SecretController until a value is sent to stopCh.
```
Comment (wattli) [inline security/pkg/pki/ca/controller/secret.go]: What value does the stopCh expect? And how to send that value?

File: security/cmd/istio_ca/main.go [hunk 5]
```diff
@@ -141,6 +142,12 @@ func runCA() {
 	stopCh := make(chan struct{})
 	sc.Run(stopCh)
 
+	serviceController := controller.NewServiceController(cs.CoreV1(), opts.namespace,
```
Comment (wattli) [inline security/cmd/istio_ca/main.go]: Can we add some comments here to explain what it does? And if possible, please do the similar thing to the above NewSecretController as well, thanks!

File: security/pkg/pki/ca/controller/secret.go (hunk 4 above)
Comment (crazytan) [reply security/pkg/pki/ca/controller/secret.go]: An empty struct would suffice `struct{}{}`. Whoever is calling the Run can send the value to `stopCh` so that the underlying controllers are stopped. See original doc https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L97:8

File: security/cmd/istio_ca/main.go (hunk 3 above)
Comment (crazytan) [reply security/cmd/istio_ca/main.go]: I felt it's better to change it because the last line is unreachable. `stopCh` passed to cache controller is actually expected to be called by `main()`, not the controller. So `<-stopCh` here causes main to block forever, but I think it's clearer to just use `select {}` here to prevent future confusion about the `stopCh` usage. See https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L97:8

File: security/cmd/istio_ca/main.go (hunk 5 above)
Comment (crazytan) [reply security/cmd/istio_ca/main.go]: Will do!

Comment (crazytan) [reply security/cmd/istio_ca/main.go]: The docs are at the struct definition for both controllers.

**Comment by wattli** [comment]
/lgtm
/approve
Thanks for doing this.
